src/lib/hosts-file.ts

```diff
@@ -0,0 +1,177 @@
+import fs from 'fs';
+import { platform } from 'os';
+import { promisify } from 'util';
+import sudo from 'sudo-prompt';
+
+const readFile = promisify( fs.readFile );
+const writeFile = promisify( fs.writeFile );
+const sudoExec = promisify( sudo.exec );
+
+// Host file paths for different operating systems
+const HOST_FILES: Record< string, string > = {
+	win32: 'C:\\Windows\\System32\\drivers\\etc\\hosts',
+	darwin: '/etc/hosts',
+	linux: '/etc/hosts',
+};
+
+// Get the appropriate hosts file path for the current platform
+const getHostsFilePath = (): string => {
+	const currentPlatform = platform();
+	const hostsPath = HOST_FILES[ currentPlatform ];
+
+	if ( ! hostsPath ) {
+		throw new Error( `Unsupported platform: ${ currentPlatform }` );
+	}
+
+	return hostsPath;
+};
+
+/**
+ * Reads the hosts file content
+ */
+export const readHostsFile = async (): Promise< string > => {
+	try {
+		const hostsPath = getHostsFilePath();
+		const content = await readFile( hostsPath, 'utf8' );
+		return content;
+	} catch ( error ) {
+		console.error( 'Error reading hosts file:', error );
+		throw error;
+	}
+};
+
+/**
+ * Writes content to the hosts file (requires elevated permissions)
+ */
+export const writeHostsFile = async ( content: string ): Promise< void > => {
+	const hostsPath = getHostsFilePath();
+	try {
+		const tempPath = '/tmp/wp-studio-hosts';
+		await writeFile( tempPath, content );
+		// @ts-expect-error promisify doesn't seem typed properly.
+		await sudoExec( `cat ${ tempPath } > ${ hostsPath }`, {
+			name: 'WordPress Studio',
+		} );
+	} catch ( error ) {
+		console.error( 'Error writing hosts file:', error );
+		throw error;
+	}
+};
+
+/**
+ * Adds a domain to the hosts file pointing to 127.0.0.1
+ */
+export const addDomainToHosts = async ( domain: string, port: number ): Promise< void > => {
+	try {
+		const hostsContent = await readHostsFile();
+
+		// Check if the domain is already in the hosts file
+		const domainRegex = new RegExp(
+			`^127\\.0\\.0\\.1\\s+${ domain.replace( /\./g, '\\.' ) }`,
+			'i'
+		);
+		if ( domainRegex.test( hostsContent ) ) {
+			return; // Domain already exists
+		}
```

Comment:
What do we do when this happens..? Display some kind of error in the app?

Reply:
To clarify, I was thinking of cases when the domain is pointing to a different IP. Edge case, perhaps, but it ties into other error handling questions like https://github.com/Automattic/studio/pull/994#discussion_r1979277381